modules/db/src/main/scala/gem/dao/UserDao.scala

```diff
@@ -11,55 +11,74 @@ import doobie.imports._
-    Statements.selectWithRoles(id, pass)
-      .list
-      .map { rows =>
-        rows.headOption.map {
-          case (i, f, l, e, s, _, _) =>
-            val map = rows.collect {
-              case (_, _, _, _, _, Some(pid), Some(role)) => Map(pid -> Set(role))
-            }.suml
-            User(i, f, l, e, s, map)
-        }
-      }
+  /**
+   * Select the root user, for system processes. Always succceds if the system is consistent;
+   * raises an exception if the root user is missing.
+   * @group Queries
+   */
+  val selectRootUser: ConnectionIO[User[ProgramRole]] =
+    OptionT(selectUser(User.Id.Root)).getOrElse(sys.error("No root user. Cannot continue."))
+
+  // Helper method to add roles to a selected user, if any.
+  private def selectUserImpl(q: Query0[User[Nothing]]): ConnectionIO[Option[User[ProgramRole]]] = {
+    for {
+      u <- OptionT(q.option)
+      r <- selectRoles(u.id).liftM[OptionT]
+    } yield u.copy(roles = r)
+  } .run
+
+  /**
+   * Select the given user, if any, with roles.
+   * @group Queries
+   */
+  def selectUser(uid: User.Id): ConnectionIO[Option[User[ProgramRole]]] =
+    selectUserImpl(Statements.selectUser(uid))
+
+  /**
+   * Select the given user, if any, with roles, if the id and password match a known user.
+   * @group Queries
+   */
+  def selectUserʹ(uid: User.Id, password: String): ConnectionIO[Option[User[ProgramRole]]] =
+    selectUserImpl(Statements.selectUserʹ(uid, password))
+
+  /**
+   * Select the map of roles associated with the given user.
+   * @group Queries
+   */
+  def selectRoles(id: User.Id): ConnectionIO[Map[Program.Id, Set[ProgramRole]]] =
+    Statements.selectRoles(id).list.map(_.foldMap { case (k, v) => Map((k -> Set(v))) })
 
+  /**
+   * Attempts to change the specified user's password, yielding success. A cause of failure (user
+   * not found, old password incorrect) is not given.
+   * @group Updates
+   */
   def changePassword(uid: User.Id, oldPassword: String, newPassword: String): ConnectionIO[Boolean] =
     Statements.changePassword(uid, oldPassword, newPassword).run.map(_ === 1)
 
   object Statements {
 
-    def select(id: String): Query0[User[Nothing]] =
-      sql"""
+    private def selectUserImpl(id: String, and: Fragment): Query0[User[Nothing]] =
+      (fr"""
         SELECT id, first, last, email, staff
         FROM gem_user
         WHERE id = $id
-      """.query[(String, String, String, String, Boolean)]
-         .map { case (i, f, l, e, s) =>
-           User(i, f, l, e, s, Map.empty)
-         }
+       """ ++ and).query[(String, String, String, String, Boolean)].map {
+        case (i, f, l, e, s) => User(i, f, l, e, s, Map.empty)
+      }
+
+    def selectUser(id: String): Query0[User[Nothing]] =
+      selectUserImpl(id, Fragment.empty)
+
+    def selectUserʹ(id: String, password: String): Query0[User[Nothing]] =
+      selectUserImpl(id, fr"AND md5 = md5($password)")
```

Comment:
If I were working on a bank I guess I would still keep nitpicking about the algorithm but I will restrain myself 😄 , so 👍 for `crypt-bf/10`.

I'd say any algorithm designed purposely designed to store passwords with the proper reputation should be good enough for us. So S2K it's perfectly fine since it's the recommended by PostgreSQL.

In any case all these functions to store passwords and its parameters should be quite easy to swap in and out at any moment so we don't need to get it perfectly right now.

Can we assume all users stored in the database will have a Google account? All Gemini personnel will have one. If that's the case, we could delegate all the authentication to Google, as @cquiroz suggested, and then avoid storing any passwords. Nowadays it's very easy to use all those authentication services, but if we need to support users without Google accounts (is there anyone in the planet without a Google account? 😄 ) then it may not be worth implementing both ways of authenticating users.